<a href="https://colab.research.google.com/github/Bertrand94/nitda-blockchain-scholarship/blob/main/curriculum/phase-2a-python/weeks-01-08-teaching/week-04-pandas-introduction/01-wednesday/exercises/week04_wed_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 4 — Pandas Introduction: Wednesday Exercises
## Phase 2a Python | PORA Academy Cohort 7

**Dataset:** `olist_orders_dataset.csv` (99,441 rows × 8 columns)

**Instructions:**
- Read each question carefully before writing code
- Run the data setup cell first — it loads `orders` into memory
- Every question has a verified expected output — check yours matches before moving on
- You may use DeepSeek if you are stuck, but try yourself for at least 5 minutes first
- Write one operation per line; do not chain multiple statements on a single line

**What you will practise today:**
- Loading a CSV with `pd.read_csv()`
- Inspecting shape, columns, dtypes, and null counts
- Filtering rows with boolean indexing and `.isin()`
- Using `.value_counts()` and `.nunique()`

## Setup — Run This First

This cell mounts your Google Drive and loads the orders dataset. Run it once before attempting any exercises.

In [8]:
import pandas as pd
from google.colab import drive
import os

drive.mount('/content/drive')
olist_path = '/content/olist_orders_dataset.csv'
orders = pd.read_csv(os.path.join(olist_path, '/content/olist_orders_dataset.csv'))
print(f"Loaded {len(orders):,} orders with {orders.shape[1]} columns")
# expected: Loaded 99,441 orders with 8 columns

Loaded 99,441 orders with 8 columns


## Question 1 — Verify the Dataset

Print the **shape** of the `orders` DataFrame and the **list of column names**.

Expected output:
```
(99441, 8)
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']
```

In [11]:
# Q1: Print orders.shape and orders.columns.tolist()
# Your code here:
# Load the Olist orders dataset
orders = pd.read_csv(os.path.join(olist_path, '/content/olist_orders_dataset.csv'))

# Rows × columns
print(orders.shape)
# expected: (99441, 8)

# Column names
print(orders.columns.tolist())
# expected: ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
#            'order_approved_at', 'order_delivered_carrier_date',
#            'order_delivered_customer_date', 'order_estimated_delivery_date']


(99441, 8)
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']


## Question 2 — Find Missing Values

Use `.isnull().sum()` to count missing values in every column.

**Then answer:** How many orders have **no delivery date** (`order_delivered_customer_date` is null)?

Expected answer: **2,965**

In [15]:
# Q2a: Print null counts for all columns
# .isnull().sum() — exact null count for every column
print(orders.isnull().sum())
print()

# Q2b: Print just the null count for order_delivered_customer_date
print(orders['order_delivered_customer_date'].isnull().sum())
# expected: 2,965


order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

2965


## Question 3 — Order Status Breakdown

Use `.value_counts()` to count how many orders have each `order_status`.

Expected output (top values):
```
delivered      96478
shipped         1107
canceled         625
unavailable      609
...
```

**Then answer:** How many unique statuses are there? Use `.nunique()`.

Expected: **8**

In [18]:
# Q3a: Print value_counts for order_status
print(orders['order_status'].value_counts())
print()

# Q3b: Print the number of unique statuses using .nunique()
print(orders['order_status'].nunique())
# expected: 8


order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

8


## Question 4 — Filter: Canceled or Unavailable Orders

Filter the `orders` DataFrame to only rows where `order_status` is either `'canceled'` **or** `'unavailable'`.

Print the number of matching rows.

Expected: **1,234** (625 canceled + 609 unavailable)

> Hint: Use `.isin()` with a list, or use `|` with two conditions (remember to wrap each in parentheses).

In [29]:
# Q4: Filter to canceled or unavailable orders and count rows
# Filter for canceled orders
canceled = orders[orders['order_status'] == 'canceled']


# Filter for unavailable orders
unavailable = orders[orders['order_status'] == 'unavailable']

# Print the number of matching rows
print(len(canceled) + len(unavailable))


1234


## Question 5 — Delivery Rate

Calculate the **percentage of all orders** that have `order_status == 'delivered'`.

Expected: **97.01%**

> Hint: Divide the count of delivered orders by the total number of orders, then multiply by 100.

In [34]:
delivered_orders_count = orders[orders['order_status'] == 'delivered'].shape[0]
total_orders_count = orders.shape[0]
delivered_order = (delivered_orders_count / total_orders_count) * 100

print(f"{delivered_order:.2f}%")
# Expected: 97.01%

97.02%


## Question 6 — Filter to Top 3 Statuses

Filter the `orders` DataFrame to only include rows where `order_status` is one of the **three most common statuses**: `'delivered'`, `'shipped'`, or `'invoiced'`.

Print the shape of the filtered DataFrame.

Expected shape: **(97899, 8)** — that is 96,478 + 1,107 + 314 rows.

> Hint: Use `.isin()` with a list of three status strings.

In [36]:
# Q6: Filter to top 3 statuses and print the shape
top_3_statuses = ['delivered', 'shipped', 'invoiced']
filtered_orders = orders[orders['order_status'].isin(top_3_statuses)]
print(filtered_orders.shape)

(97899, 8)


## Question 7 — Healthy Orders Only

The operations team defines a "healthy" order as one that is **not** `'canceled'` and **not** `'unavailable'`.

Filter `orders` to only healthy orders using the `~` (NOT) operator with `.isin()`.

Print the count of healthy orders.

Expected: **98,207** (99,441 − 625 − 609)

In [37]:
# Q7: Filter to healthy orders using ~ with .isin() and print the count
# given
problem_statuses = ['canceled', 'unavailable']
healthy_orders = orders[~orders['order_status'].isin(problem_statuses)]
print(len(healthy_orders))

98207


## Group Exercise — Olist Orders Analysis

Work with your group. Attempt each task independently first (5 minutes), then compare answers. Use DeepSeek only after a genuine attempt.

**Verify every answer against the expected value before marking it done.**

Using the `orders` DataFrame already loaded above:

In [ ]:
# ── Group Task 1 ─────────────────────────────────────────────
# Verify: orders.shape == (99441, 8)
# Your code here:

# ── Group Task 2 ─────────────────────────────────────────────
# How many orders have NO delivery date?
# orders['order_delivered_customer_date'].isnull().sum()
# Expected: 2,965
# Your code here:

# ── Group Task 3 ─────────────────────────────────────────────
# Filter to 'canceled' OR 'unavailable' orders. How many rows?
# Expected: 1,234
# Your code here:

# ── Group Task 4 ─────────────────────────────────────────────
# What % of all orders are 'delivered'?
# Expected: 97.01%
# Your code here:

# ── Group Task 5 ─────────────────────────────────────────────
# Filter to only the 3 most common statuses: delivered, shipped, invoiced
# Hint: use isin(['delivered', 'shipped', 'invoiced'])
# Expected row count: 97,899
# Your code here:


## Group Task 1 — Verify the Dataset

Print the **shape** of the `orders` DataFrame and the **list of column names**.

Expected output:
```
(99441, 8)
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

In [39]:
import pandas as pd
from google.colab import drive
import os

drive.mount('/content/drive')
olist_path = '/content/olist_orders_dataset.csv'
orders = pd.read_csv(os.path.join(olist_path, '/content/olist_orders_dataset.csv'))
print(f"Loaded {len(orders):,} orders with {orders.shape[1]} columns")

Loaded 99,441 orders with 8 columns


In [40]:
# Print orders.shape and orders.columns.tolist()
# Load the Olist orders dataset
orders = pd.read_csv(os.path.join(olist_path, '/content/olist_orders_dataset.csv'))

# Rows × columns
print(orders.shape)
# expected: (99441, 8)

# Column names
print(orders.columns.tolist())
# expected: ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
#            'order_approved_at', 'order_delivered_carrier_date',
#            'order_delivered_customer_date', 'order_estimated_delivery_date']


(99441, 8)
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']


## Group Task 2 — Find Missing Values

Use `.isnull().sum()` to count missing values in every column.

**Then answer:** How many orders have **no delivery date** (`order_delivered_customer_date` is null)?

Expected answer: **2,965**

In [42]:
# How many orders have NO delivery date?
orders['order_delivered_customer_date'].isnull().sum()
print(orders['order_delivered_customer_date'].isnull().sum())

2965


## Group Task 3 — Filter: Canceled or Unavailable Orders

Filter the `orders` DataFrame to only rows where `order_status` is either `'canceled'` **or** `'unavailable'`.

Print the number of matching rows.

Expected: **1,234** (625 canceled + 609 unavailable)

> Hint: Use `.isin()` with a list, or use `|` with two conditions (remember to wrap each in parentheses).

In [44]:
# Filter to 'canceled' OR 'unavailable' orders. How many rows?
canceled = orders[orders['order_status'] == 'canceled']
unavailable = orders[orders['order_status'] == 'unavailable']

# Print the number of matching rows
print(len(canceled) + len(unavailable))


1234


## Group Task 4 — Delivery Rate

Calculate the **percentage of all orders** that have `order_status == 'delivered'`.

Expected: **97.01%**

> Hint: Divide the count of delivered orders by the total number of orders, then multiply by 100.

In [45]:
# What % of all orders are 'delivered'?
delivered_orders_count = orders[orders['order_status'] == 'delivered'].shape[0]
total_orders_count = orders.shape[0]
delivered_order = (delivered_orders_count / total_orders_count) * 100

print(f"{delivered_order:.2f}%")

97.02%


## Group Task 5 — Healthy Orders Only

The operations team defines a "healthy" order as one that is **not** `'canceled'` and **not** `'unavailable'`.

Filter `orders` to only healthy orders using the `~` (NOT) operator with `.isin()`.

Print the count of healthy orders.

Expected: **98,207** (99,441 − 625 − 609)

In [46]:
# Filter to only the 3 most common statuses: delivered, shipped, invoiced
top_3_statuses = ['delivered', 'shipped', 'invoiced']
filtered_orders = orders[orders['order_status'].isin(top_3_statuses)]
print(filtered_orders.shape)

(97899, 8)
